lib/storage/authgateway_test.go

```diff
@@ -0,0 +1,244 @@
+/*
+Copyright 2019 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package storage
+
+import (
+	"time"
+
+	"github.com/gravitational/gravity/lib/compare"
+	teleconfig "github.com/gravitational/teleport/lib/config"
+	teleservices "github.com/gravitational/teleport/lib/services"
+	teleutils "github.com/gravitational/teleport/lib/utils"
+	check "gopkg.in/check.v1"
+)
+
+type AuthGatewaySuite struct{}
+
+var _ = check.Suite(&AuthGatewaySuite{})
+
+func (s *AuthGatewaySuite) TestResourceParsing(c *check.C) {
+	spec := `kind: authgateway
+version: v2
+spec:
+  connection_limits:
+    max_connections: 2000
+    max_users: 20
+  authentication:
+    type: oidc
+    second_factor: "off"
+    connector_name: google
+  client_idle_timeout: 60s
+  disconnect_expired_cert: true
+  public_addr: ["example.com"]
+  ssh_public_addr: ["ssh.example.com"]
+  kubernetes_public_addr: ["k8s.example.com"]
+  web_public_addr: ["web.example.com"]
+`
+	gw, err := UnmarshalAuthGateway([]byte(spec))
+	c.Assert(err, check.IsNil)
+	c.Assert(gw, compare.DeepEquals, NewAuthGateway(AuthGatewaySpecV2{
+		ConnectionLimits: &ConnectionLimits{
+			MaxConnections: int64p(2000),
+			MaxUsers:       intp(20),
+		},
+		Authentication: &teleservices.AuthPreferenceSpecV2{
+			Type:          "oidc",
+			SecondFactor:  "off",
+			ConnectorName: "google",
+		},
+		ClientIdleTimeout:     durp(60 * time.Second),
+		DisconnectExpiredCert: teleservices.NewBoolOption(true),
+		PublicAddr:            &[]string{"example.com"},
```

Comment:
To be honest, I don't the benefit to using pointers to slice.

Reply:
It's for the same reason as I mentioned above: to be able to tell whether it's an empty slice or the field wasn't set.

Reply:
Well, you can use `nil` for this - empty slice and `nil` slice are different.